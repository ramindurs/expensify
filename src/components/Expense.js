import React from 'react';
import {Link} from 'react-router-dom';
import moment from "moment";
import numeral from 'numeral';

export const Expense = ({ id, description, amount, createdAt, dispatch, history}) => {
    return (
        <div>
            {id} <Link to={`/edit/${id}` }>{description}</Link>
            <p>
                {numeral(amount / 100).format('0,0.00')}
                -
                {moment(createdAt).format('MMM Do YYYY')}
            </p>
        </div>
    );
};

export default Expense;
