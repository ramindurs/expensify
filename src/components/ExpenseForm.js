import 'react-dates/initialize';
import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from "react-dates";


export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        const expense = props.expenseToEdit;
        this.state = {
            description: expense && expense.description ? expense.description : '',
            note: expense && expense.note ? expense.note : '',
            amount: expense && expense.amount ? this.convertIntToFloatString(expense.amount) : '',
            createdAt: expense && expense.createdAt ? moment(expense.createdAt) : moment(),
            calendarFocus: false,
            error: ''
        };
    }
    convertIntToFloatString = (intnum) => parseFloat(intnum) / 100;

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({description}));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({amount}));
        }
    };
    onDateChange = (date) => {
        if (!!date) {
            this.setState(() => ({createdAt: date}));
        }
    };
    onFocusChange = ({focused}) => {
        this.setState(() => ({calendarFocus: focused}));
    };
    onSubmitForm = (e) => {
        e.preventDefault();
        if (this.state.description && this.state.amount) {
            this.setState(() => ({error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                note: this.state.note,
                amount: Math.round(parseFloat(this.state.amount, 10) * 100),
                createdAt: this.state.createdAt.valueOf()
            });
        } else {
            this.setState(() => ({error: 'Please provide description and amount'}));
        }
    };

    render() {
        return (
            <div>
                {this.state.error ? <p>{this.state.error}</p> : ''}
                <form onSubmit={this.onSubmitForm}>
                    <input
                        type="text"
                        placeholder={'description'}
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                        autoFocus/>
                    <input
                        type="text"
                        placeholder={'amount'}
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocus}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        onChange={this.onNoteChange}
                        value={this.state.note}
                        placeholder={'Add a note'}>
                    </textarea>
                    <button>{this.props.submitMessage}</button>
                </form>
            </div>
        )
    };
};
