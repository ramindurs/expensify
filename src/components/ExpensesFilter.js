import React from 'react';
import {connect} from 'react-redux';
import { DateRangePicker} from "react-dates";
import {setSortByAmount, setSortByDate, setTextFilter, setStartDate, setEndDate} from "../actions/filters";

export class ExpensesFilter extends React.Component {
    state = {
        calendarFocused: null
    };

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onTextFilterChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    setSortBy = (e) => {
        const sortBy = e.target.value;
        switch (sortBy) {
            case 'amount':
                this.props.setSortByAmount();
                return;
            default:
                this.props.setSortByDate();
                return;
        }
    };

    onFocusChange = (focusedInput) => {
        this.setState(() => ({ calendarFocused: focusedInput}));
    };

    render() {
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={this.onTextFilterChange}/>
                <select
                    value={this.props.filters.sortBy}
                    onChange={this.setSortBy}>
                    <option value={'date'}>Date</option>
                    <option value={'amount'}>Amount</option>
                </select>
                <DateRangePicker
                    startDateId={'startDate'}
                    startDate={this.props.filters.startDate}
                    endDateId={'endDate'}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                />
            </div>
        );
    }
}

const mappedStateToProps = (state) => {
    return {
        filters: state.filters
    }
};

const mappedDispatchToProps = (dispatch) => {
    return {
        setStartDate: (date) => dispatch(setStartDate(date)),
        setEndDate: (date) => dispatch(setEndDate(date)),
        setTextFilter: (text) => dispatch(setTextFilter(text)),
        setSortByAmount: () => dispatch(setSortByAmount()),
        setSortByDate: () => dispatch(setSortByDate())
    };
};

export default connect(mappedStateToProps, mappedDispatchToProps)(ExpensesFilter);