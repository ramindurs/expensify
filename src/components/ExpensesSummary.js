import React from 'react';
import {connect} from "react-redux";
import selectExpenses from '../selectors/expenses';
import expensesTotal from "../selectors/expenses-total";

export class ExpensesSummary extends React.Component {
    getTotalNumber = () => (this.props.expenses.length);
    getTotalAmount = () => (this.props.expensesTotal(this.props.expenses));

    render() {
        return (
            <div>
                <h4>Expenses Summary</h4>
                {this.getTotalNumber() === 0 ? 'Number of Expenses: 0' :
                    'Number of Expenses: ' + this.getTotalNumber() + ' Total Amount: Â£' + this.getTotalAmount()
                }
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters),
        expensesTotal: (filteredExpenses) => (expensesTotal(filteredExpenses) / 100)
    }
};

export default connect(mapStateToProps)(ExpensesSummary);
