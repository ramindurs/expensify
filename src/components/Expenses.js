import React from 'react';
import {connect} from "react-redux";
import Expense from "./Expense";
import selectExpenses from '../selectors/expenses';

export const Expenses = (props) => (
    <div>
        <h1>Expenses List</h1>
        {props.expenses.map((expense, index) =>
            <Expense key={index}
                     {...expense}
            />)}
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
};

export default connect(mapStateToProps)(Expenses);
