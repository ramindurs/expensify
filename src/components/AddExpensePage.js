import React from 'react';
import {connect} from 'react-redux';
import {startAddExpense} from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push('/dashboard');
    };

    render() {
        return (
            <div>
                <h3>Add Expenses</h3>
                <ExpenseForm
                    submitMessage={'Add Expense'}
                    onSubmit={this.onSubmit}/>
            </div>
        );
    }
}

const mapDispatchTpProps = (dispatch) => {
    return {
        startAddExpense: (expense) => dispatch(startAddExpense(expense))
    };
};

export default connect(undefined, mapDispatchTpProps)(AddExpensePage);
