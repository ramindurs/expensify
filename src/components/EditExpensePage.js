import React from 'react';
import {connect} from "react-redux";
import ExpenseForm from "./ExpenseForm";
import {startEditExpense, startRemoveExpense} from "../actions/expenses";

export class EditExpensePage extends React.Component {
    removeExpense = (id) => {
        this.props.startRemoveExpense(id);
        this.props.history.push('/dashboard');
    };
    editExpense = (id, expense) => {
        this.props.startEditExpense(id, expense);
        this.props.history.push('/dashboard');
    };

    render() {
        return (
            <div>
                <h3>Edit Expense ID {this.props.expense.id}</h3>
                <ExpenseForm
                    expenseToEdit={this.props.expense}
                    submitMessage={'Update Expense'}
                    onSubmit={(expense) => {
                        this.editExpense(this.props.expense.id, expense)
                    }}
                />
                <button onClick={() => {
                    this.removeExpense(this.props.expense.id)
                }
                }>Delete
                </button>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find(e => e.id === props.match.params.id)
    };
};

const matchDispatchToProps = (dispatch, props) => {
    return {
        startEditExpense: (id, expense) => dispatch(startEditExpense({
            id: id,
            updates: expense
        })),
        startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
    };
};

export default connect(mapStateToProps, matchDispatchToProps)(EditExpensePage);
