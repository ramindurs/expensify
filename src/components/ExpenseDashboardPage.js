import React from 'react';
import Expenses from "./Expenses";
import ExpensesFilter from "./ExpensesFilter";
import ExpensesSummary from "./ExpensesSummary";

const ExpenseDashboardPage = () => (
    <div>
        <ExpensesFilter/>
        <ExpensesSummary/>
        <Expenses/>
    </div>
);

export default ExpenseDashboardPage;