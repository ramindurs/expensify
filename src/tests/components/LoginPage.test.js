import React from 'react';
import {shallow} from 'enzyme';
import {LoginPage} from "../../components/LoginPage";

test('should render login page', () => {
    const wrapper = shallow(<LoginPage startLogin={() => {
    }}/>);

    expect(wrapper).toMatchSnapshot();
});

test('should call startLogin when login button is clicked', () => {
    const startLoginSpy = jest.fn();
    const wrapper = shallow(<LoginPage startLogin={startLoginSpy}/>);

    wrapper.find('button').simulate('click');

    expect(startLoginSpy).toHaveBeenCalled();
});

