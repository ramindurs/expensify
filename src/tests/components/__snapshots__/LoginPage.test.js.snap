// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render login page 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
